capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -202,9 +450,479 @@ impl<'a> NonvolatileStorage<'a> {
+            header_buffer: TakeCell::new(header_buffer),
+            next_unallocated_region_header_address: OptionalCell::empty(),
         }
     }
 
+    // Capsule-level initialization that verifies the magic header,
+    // corrects it if needed
+    pub fn init(&self) -> Result<(), ErrorCode> {
+        self.check_magic_header()
+    }
+
+    // App-level initialization that allocates a region for an app or fetches
+    // an app's existing region from nonvolatile storage
+    fn init_app(&self, processid: ProcessId) -> Result<(), ErrorCode> {
+        // Signal that this app requested a storage region. If it isn't
+        // allocated immediately, it will be handled after previous requests
+        // are handled.
+        self.apps.enter(processid, |app, _kernel_data| {
+            app.has_requested_region = true;
+        })?;
+
+        // Start traversing the storage regions to find where the requesting app's
+        // storage region is located. If it doesn't exist, a new one will be allocated
+        self.start_region_traversal()
+    }
+
+    // Start reading app region headers. The first read will be from the region immediately
+    // following the magic header. See the storage layout diagram at the top of this file.
+    fn start_region_traversal(&self) -> Result<(), ErrorCode> {
+        let first_header_address = self.userspace_start_address + MAGIC_HEADER_LEN;
+        self.read_region_header(first_header_address)
+    }
+
+    // Find an app that previously requested a nonvolatile
+    // region and doesn't have one assigned.
+    fn find_app_to_allocate_region(&self) -> Option<ProcessId> {
+        for app in self.apps.iter() {
+            let processid = app.processid();
+            // find the first app that needs to be allocated
+            let app_to_allocate = app.enter(|app, _kernel_data| {
+                // if the app previously requested a region and
+                // hasn't been allocated one yet
+                if app.has_requested_region && app.region.is_none() {
+                    Some(processid)
+                } else {
+                    // no apps need to be allocated
+                    None
+                }
+            });
+
+            if app_to_allocate.is_some() {
+                return app_to_allocate;
+            }
+        }
+        None
+    }
+
+    fn allocate_app_region(&self, processid: ProcessId) -> Result<(), ErrorCode> {
+        // can't allocate a region if we haven't previously traversed existing regions
+        // and found where they stop
+        let new_header_addr = self.next_unallocated_region_header_address.get().ok_or(ErrorCode::FAIL)?;
+
+        // Apps must have a fixed ShortID value. LocallyUnique apps
+        // are not allowed since we need a unique fixed ID to write to
+        // storage and use to identify apps.
+        let ShortId::Fixed(shortid) = processid.short_app_id() else {
+            return Err(ErrorCode::NOSUPPORT);
+        };
```

Comment:
Oh I should have caught this earlier. Because this is nonvolatile storage, apps must have permission to use it. We need to use this (https://github.com/tock/tock/blob/master/kernel/src/storage_permissions.rs#L210) API, not `.short_app_id()`.

Reply:
And then likewise we need to use the read permissions for reads, etc.

Reply:
Should we store the `write_id` to storage in the region header instead of the `ShortID::Fixed` value? Or just use this API for checking permissions before allocating region and read/writes?

Reply:
> Should we store the `write_id` to storage in the region header instead of the `ShortID::Fixed` value?

Yes.

(Of course it's the same, but the code will make more sense just using the storage permissions interface.)

Reply:
Updated in this commit https://github.com/tock/tock/pull/4109/commits/002e08f4b5e52f08881759cf3c88c7578adde1c8.

If you could look over it, that would be great to make sure I'm using the API as intended.

However, with these changes my testing apps aren't able to allocate regions due to failing the new permission checks. I was looking over the TRD for Storage Permissions over here https://book.tockos.org/trd/trd-storage-permissions.html?highlight=permissions#7-specifying-permissions. 

So I need to implement a `PermissionsPolicy` trait in the board `main.rs` which chooses how to hand out app storage permissions?